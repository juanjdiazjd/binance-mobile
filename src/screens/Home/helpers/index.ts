export const mountToSatoshi = (mount: number): string =>
  (mount / 100000000).toFixed(8);
