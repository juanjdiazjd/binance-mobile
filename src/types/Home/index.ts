export enum ItemsHomeStackList {
  Home = 'Home',
  Exchange = 'Exchange',
  OrderBook = 'OrderBook',
}
