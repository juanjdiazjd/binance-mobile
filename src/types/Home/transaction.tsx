export enum Status {
  completed = 'completed',
  failed = 'failed',
}
export type Transaction = {
  id: string;
  date: string;
  mount: string;
  address: string;
  status: Status;
};
export type Transactions = {
  Transactions: Transaction[];
  ErrorMsg: string;
};
