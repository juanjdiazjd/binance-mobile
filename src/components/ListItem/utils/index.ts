import theme from '../../../theme';
import {Status} from '../../../types/Home/transaction';
const dayjs = require('dayjs');

export const colorByStatus = (status?: Status): string => {
  switch (status) {
    case Status.completed:
      return theme.colors.green;
    default:
      return theme.colors.red;
  }
};

export const parseDate = (date: string, complete: boolean): string => {
  return complete
    ? dayjs(date).format('DD/MM/YYYY hh:mm:ss')
    : dayjs(date).format('DD/MM/YYYY');
};
