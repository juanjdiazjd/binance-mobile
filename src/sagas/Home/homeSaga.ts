import {takeLatest} from '@redux-saga/core/effects';
import {HomeActionTypes} from '../../actions/actionTypes';
import {PERMISSIONS, request} from 'react-native-permissions';
import {BUE_LAT_LONG} from '../../utils/locationCustom';
import Geolocation from 'react-native-geolocation-service';
import {actionType} from '../../types/actionType';
import {call, put} from 'redux-saga/effects';
import weatherApi from '../../api';
import {HomeActions} from '../../actions/homeActions';
import {LatLong} from '../../types/Home/weather';

// Constants
const CONFIG_DEFAULT_LOCATION = {
  enableHighAccuracy: true,
  timeout: 15000,
  maximumAge: 10000,
};

//Types
export type ConfigPosition = {
  enableHighAccuracy: boolean;
  timeout: number;
  maximumAge: number;
};

//API Services
const {weatherServices: weatherAPI} = weatherApi;

//Methods
function* fetchCurrentWeather() {
  try {
    const {coords: coords}: Geolocation.GeoPosition =
      yield currentPositionPromise(CONFIG_DEFAULT_LOCATION);
    const bodyRequest: LatLong = {lat: coords.latitude, lon: coords.longitude};
    const {ok, problem, data} = yield call(
      weatherAPI.currentWeather,
      bodyRequest,
    );
    yield put(HomeActions.IsFetchingCurrentMethod(true));

    if (!ok) {
      yield put(HomeActions.IsFetchingCurrentMethod(false));
      throw new Error(problem);
    }
    yield put(HomeActions.IsFetchingCurrentMethod(false));
    yield put(HomeActions.setCurrentWeather(data));
  } catch (error) {}
}
function* fetchForecastWeather(action:actionType<LatLong>) {
  try {
    const {coords: coords}:  Geolocation.GeoPosition =
      yield currentPositionPromise(CONFIG_DEFAULT_LOCATION);
    const bodyRequest: LatLong = {lat: action.payload?.lat || coords.latitude, lon: action.payload?.lon || coords.longitude};
    const {ok, problem, data} = yield call(
      weatherAPI.forecastWeather,
      bodyRequest,
    );
    yield put(HomeActions.IsFetchingForecastMethod(true));

    if (!ok) {
      yield put(HomeActions.IsFetchingForecastMethod(false));
      //Falta setear error
      throw new Error(problem);
    }
    yield put(HomeActions.IsFetchingForecastMethod(false));
    if(action.payload?.lat && action.payload?.lon){
      yield put(HomeActions.setForecastSearchWeather(data));

    }else{
      yield put(HomeActions.setForecastWeather(data));

    }
  } catch (error) {}
}

export const currentPositionPromise: (
  config: ConfigPosition,
) => Promise<Geolocation.GeoPosition> = async () =>
  new Promise(async (resolve, reject) => {
    request(PERMISSIONS.IOS.LOCATION_ALWAYS).then(result => {
      if (result == 'granted') {
        Geolocation.getCurrentPosition(
          currentPositionResponse => {
            resolve(currentPositionResponse);
          },
          error => {
            if (error.code == 1) {
              resolve({
                coords: {
                  accuracy: 5,
                  altitude: 0,
                  altitudeAccuracy: -1,
                  heading: -1,
                  latitude: BUE_LAT_LONG.lat,
                  longitude: BUE_LAT_LONG.long,
                  speed: -1,
                },
                timestamp: 1632790198315.311,
              });
            }
          },
          CONFIG_DEFAULT_LOCATION,
        );
      }
    });
    request(PERMISSIONS.ANDROID.ACCESS_FINE_LOCATION).then(result => {
      if (result == 'granted') {
        Geolocation.getCurrentPosition(
          currentPositionResponse => {
            resolve(currentPositionResponse);
          },
          error => {},
          CONFIG_DEFAULT_LOCATION,
        );
      }
    });
  });

function* homeSagas() {
  yield takeLatest(HomeActionTypes.GetCurrentWeather, fetchCurrentWeather);
  yield takeLatest(HomeActionTypes.GetForecastWeather, fetchForecastWeather);

}

export default homeSagas;
