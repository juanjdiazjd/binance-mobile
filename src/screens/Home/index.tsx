import * as React from 'react';
import {Dimensions, ScrollView, StatusBar, View} from 'react-native';
import {WrapperView} from '../../components/Wrappers/SafeAreaWrapper';
import {ContentView} from '../../components/Wrappers/ContentView';
import {TextType} from '../../components/Text/TextView';
import {strings} from './strings';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import styled from 'styled-components';
import {RootStackParamList} from '../../utils/constants';
import {Header} from '../../components/UI/Header';
import apiServices from '../../api';
import useSWR from 'swr';
import {Logo} from '../History';
import {feeTypeById, getBalanceInFiat} from './utils';
import CurrencyInput from 'react-native-currency-input';
import {useMountWithFee} from '../../hooks/use-mount-with-fee';
import {FeeTypes, TransactionStatus} from '../../types/responseType';
import theme from '../../theme';
import CheckBoxFee from '../../components/CheckBoxFee';
import {ICheckboxButton} from 'react-native-bouncy-checkbox-group';
import {SeparatorView} from '../../components/UI/SeparatorView';
import {TextCustom} from '../../components/ListItem/ListItem';
import {useState} from 'react';
import {Modal, Portal, Text, Button, Provider} from 'react-native-paper';
import {Wallet} from '../../components/Wallet';

const {balanceServices: balanceAPI, ripioServices: ratesAPI} = apiServices;

const MIN_VALUE = 0.0;
const PRECISION_VALUE = 8;
const windowWidth = Dimensions.get('window').width;
const windowHeight = Dimensions.get('window').height;

export const ContainerBalance = styled(View)`
  align-self: center;
  height: 40px;
  width: 90%;
  padding: 5px;
  border-radius: 5px;
  background-color: ${theme.colors.secondary};
`;
const ContainerCurrencyInput = styled(View)`
  align-items: center;
  height: 40px;
  padding: 5px;
  border-radius: 5px;
  background-color: ${theme.colors.background};
`;

const ContainerButton = styled(View)`
  flex-direction: row;
  justify-content: center;
`;
const CustomButton = styled(Button)`
  justify-content: center;
  margin-top: 30px;
`;

const HomeScreen = ({}: NativeStackScreenProps<RootStackParamList, 'Home'>) => {
  const [value, setValue] = useState(0);
  const [feeType, setFeeType] = useState<FeeTypes>(FeeTypes.hourFee);
  const [isModalVisible, setModalVisible] = useState(false);

  const {data: balance, mutate: mutateGet} = useSWR(
    'getBalance',
    balanceAPI.getBalance,
  );
  const myBalance = balance?.Balance.balance ?? 0;
  const {data: rates} = useSWR('getRates', ratesAPI.getRates);
  const [{mountWithFee, availableBalance}] = useMountWithFee(
    myBalance,
    value,
    feeType,
  );

  const onChangeValue = (mount: number) => {
    setValue(mount);
  };

  const onHandleSelectedCheckBox = (item: ICheckboxButton) => {
    setFeeType(feeTypeById(item.id));
  };

  const toggleModal = (dismiss?: boolean) => {
    setModalVisible(!isModalVisible);
    if (dismiss) {
      mutateGet();
    }
  };
  return (
    <WrapperView>
      <ContentView fullWidth={true}>
        <ScrollView
          contentContainerStyle={styles.scrollView}
          keyboardShouldPersistTaps="handled">
          <StatusBar barStyle="dark-content" />
          <Header
            title={strings.home.infoTitle}
            subtitle={strings.home.infoSubtitle}
            textTypeTitle={TextType.bigBoldLato}
            buttonBack={false}
            secondaryComponent={() => (
              <View>
                <Logo source={require('../../assets/ripio-logo.png')} />
                <ContainerBalance>
                  <Text>{`${strings.home.balanceInBtc}: ${myBalance}`}</Text>
                  <Text>
                    {rates?.rates.ARS_SELL &&
                      `${strings.home.ARS}: ${getBalanceInFiat(
                        myBalance,
                        rates?.rates.ARS_SELL,
                      )}`}
                  </Text>
                </ContainerBalance>
              </View>
            )}
          />
          <ContainerCurrencyInput>
            <CurrencyInput
              value={value}
              style={styles.currencyInput}
              onChangeValue={onChangeValue}
              placeholder={'0 BTC'}
              delimiter="."
              separator=","
              precision={PRECISION_VALUE}
              minValue={MIN_VALUE}
            />
            <TextCustom>
              {`${strings.home.mountTotal} ${mountWithFee}`}
            </TextCustom>
            {!availableBalance && (
              <TextCustom color={theme.colors.red}>
                {strings.home.notBalance}
              </TextCustom>
            )}
          </ContainerCurrencyInput>
          <SeparatorView width={'40px'} height={'40px'} />
          <CheckBoxFee onHandlePress={onHandleSelectedCheckBox} />
          <ContainerButton>
            <CustomButton
              disabled={!availableBalance || !value}
              icon={'send'}
              mode="contained"
              onPress={() => toggleModal(false)}>
              {strings.home.send}
            </CustomButton>
          </ContainerButton>
          <Provider>
            <Portal>
              <Modal
                visible={isModalVisible}
                onDismiss={() => toggleModal(true)}
                contentContainerStyle={styles.contentModalStyle}>
                <Wallet
                  totalBalance={myBalance}
                  mountWithFee={mountWithFee}
                  refetch={() => toggleModal(true)}
                  status={
                    value === 0.00000001
                      ? TransactionStatus.failed
                      : TransactionStatus.completed
                  }
                />
              </Modal>
            </Portal>
          </Provider>
        </ScrollView>
      </ContentView>
    </WrapperView>
  );
};
const styles = {
  scrollView: {flexGrow: 1},
  currencyInput: {fontSize: 30},
  contentModalStyle: {
    backgroundColor: theme.colors.background,
    width: windowWidth,
    height: windowHeight / 2,
  },
};
export default HomeScreen;
