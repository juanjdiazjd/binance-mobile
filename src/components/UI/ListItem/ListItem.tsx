import * as React from 'react';
import {Text, TouchableOpacity, View, ViewProps} from 'react-native';
import styled from 'styled-components';
import theme from '../../../theme';
import {Status, Transaction} from '../../../types/Home/transaction';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import {colorByStatus, parseDate} from './utils';

const ListContainer = styled(TouchableOpacity)`
  background-color: ${theme.colors.white};
  height: auto;
  width: 350px;
  padding: 15px;
  margin-bottom: 10px;
  border-radius: 10px;
  flex-direction: row;
  justify-content: space-between;
`;

const ComponentContainer = styled(View)`
  flex-direction: row;
  justify-content: center;
  height: auto;
  width: auto;
`;

export const TextMount = styled(Text)`
  color: black;
  align-self: flex-end;
  font-size: 16px;
`;

export const TextCustom = styled(Text)`
  top: 10%;
  padding-left: 20px;
  font-size: 14px;
`;

export const TextDate = styled(Text)`
  top: 10%;
  color: ${theme.colors.gray};
  padding-left: 20px;
  font-size: 14px;
`;

export const Column = styled(View)`
  flex-direction: column;
`;

export const Row = styled(View)`
  flex-direction: row;
`;
const CustomIcon = styled(Icon)<{
  status: Status;
}>`
  position: absolute;
  bottom: -15px;
  right: -15px;
  color: ${({status}) => colorByStatus(status)};
  width: 30px;
  height: 30px;
  padding-left: 2px;
  margin-top: 10px;
`;

interface ListItemViewProps extends ViewProps {
  item: Transaction;
  onHandlePress: (item: Transaction) => void;
}

export const ListItem: React.FunctionComponent<ListItemViewProps> = ({
  item,
  onHandlePress,
}) => {
  return (
    <ComponentContainer>
      <ListContainer onPress={() => onHandlePress(item)}>
        <Row>
          <Column>
            <Icon name="bitcoin" size={40} color={theme.colors.secondary} />
            <CustomIcon
              status={item.status}
              name={
                item.status === Status.completed
                  ? 'check-circle'
                  : 'close-circle'
              }
              size={20}
              color={colorByStatus(item.status)}
            />
          </Column>
          <Column>
            <TextCustom>Env√≠o</TextCustom>
            <TextDate>{parseDate(item.date, false)}</TextDate>
          </Column>
        </Row>
        <TextMount>$ {item.mount}</TextMount>
      </ListContainer>
    </ComponentContainer>
  );
};

export default ListItem;
