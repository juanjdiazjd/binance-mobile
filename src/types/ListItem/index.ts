import {ItemsHomeStackList} from 'types/Home';

export type Feature = {
  id: number;
  image: string;
  title: string;
  screenName: ItemsHomeStackList;
  color: string;
  backgroundColor: string;
};
