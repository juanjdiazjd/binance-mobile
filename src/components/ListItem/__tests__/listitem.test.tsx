import React from 'react';
import renderer from 'react-test-renderer';
import {Status, Transaction} from '../../../types/Home/transaction';
import ListItem from '../ListItem';

jest.mock('swr');
jest.mock('react-native-paper');
jest.mock('react-native-localization');
jest.resetAllMocks();
jest.useFakeTimers();

const MOCK_DATA_LIST_ITEM: Transaction = {
  id: '1ac74e7e-e65f-44a9-be15-9015019e3e2f',
  date: '2022-07-11T16:58:24.821970868Z',
  mount: '0.0002',
  address: '0xqeqweqeq',
  status: Status.completed,
};

const ListItemComponent = (): JSX.Element => (
  <ListItem item={MOCK_DATA_LIST_ITEM} onHandlePress={jest.fn()} />
);

it('Should listi tem component', () => {
  const tree = renderer.create(ListItemComponent()).toJSON();
  expect(tree).toMatchSnapshot();
});
