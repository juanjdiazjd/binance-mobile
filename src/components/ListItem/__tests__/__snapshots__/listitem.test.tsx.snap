// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should listi tem component 1`] = `
<View
  className="sc-gsnTZi fluHiX"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      className="sc-gKXOVf biUkuu"
    >
      <View
        className="sc-jSMfEi fsKpmn"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#FFB400",
                "fontSize": 40,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰠓
        </Text>
        <Text
          allowFontScaling={false}
          className="sc-iBkjds jfEoOc"
          selectable={false}
          status="completed"
          style={
            Array [
              Object {
                "color": "#27ae60",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰗠
        </Text>
      </View>
      <View
        className="sc-jSMfEi fsKpmn"
      >
        <Text
          className="sc-hKMtZM hYUqiH"
        >
          Envío
        </Text>
        <Text
          className="sc-eCYdqJ lkOUjh"
        >
          11/07/2022
        </Text>
      </View>
    </View>
    <Text
      className="sc-dkzDqf izhazh"
    >
      $ 
      0.0002
    </Text>
  </View>
</View>
`;
